src/coreclr/vm/jitinterface.cpp

```diff
@@ -11667,43 +11662,61 @@ bool CEEInfo::getStaticFieldContent(CORINFO_FIELD_HANDLE fieldHnd, uint8_t* buff
+                        unsigned numSlots = (structType.GetSize() + TARGET_POINTER_SIZE - 1) / TARGET_POINTER_SIZE;
+                        CQuickBytes gcPtrs;
+                        BYTE* ptr = static_cast<BYTE*>(gcPtrs.AllocThrows(numSlots));
+                        CEEInfo::getClassGClayoutStatic(structType, ptr);
+
+                        _ASSERT(numSlots > 0);
+
+                        bool seenGcSlot = false;
+                        bool canUseMemcpy = true;
+                        for (unsigned i = 0; i < numSlots; i++)
                         {
-                            // TODO: If subField is also a struct we might want to inspect its fields too
-                            if (subField->GetOffset() == (DWORD)valueOffset && subField->IsObjRef())
+                            if (ptr[i] == TYPE_GC_NONE)
                             {
-                                GCX_COOP();
+                                // Not a GC slot
+                                continue;
+                            }
+                            seenGcSlot = true;
 
-                                // Read field's value
-                                Object* subFieldValue = nullptr;
-                                memcpy(&subFieldValue, (uint8_t*)baseAddr + valueOffset, bufferSize);
+                            const unsigned gcSlotBegin = i * TARGET_POINTER_SIZE;
+                            const unsigned gcSlotEnd = gcSlotBegin + TARGET_POINTER_SIZE;
 
-                                if (subFieldValue == nullptr)
-                                {
-                                    // Report null
-                                    memset(buffer, 0, bufferSize);
-                                    result = true;
-                                }
-                                else if (GCHeapUtilities::GetGCHeap()->IsInFrozenSegment(subFieldValue))
+                            if (gcSlotBegin >= (unsigned)valueOffset && gcSlotEnd <= (unsigned)(valueOffset + bufferSize))
+                            {
+                                // GC slot intersects with our valueOffset + bufferSize - we can't use memcpy...
+                                canUseMemcpy = false;
+
+                                // ...unless we're interested in that GC slot's value itself
+                                if (gcSlotBegin == (unsigned)valueOffset && gcSlotEnd == (unsigned)(valueOffset + bufferSize) && ptr[i] == TYPE_GC_REF)
                                 {
-                                    CORINFO_OBJECT_HANDLE handle = getJitHandleForObject(
-                                        ObjectToOBJECTREF(subFieldValue), /*knownFrozen*/ true);
+                                    Object* gcSlotValue = nullptr;
+                                    memcpy(&gcSlotValue, (uint8_t*)baseAddr + gcSlotBegin, sizeof(CORINFO_OBJECT_HANDLE));
```

Comment:
```suggestion
                                    memcpy(&gcSlotValue, (uint8_t*)baseAddr + gcSlotBegin, sizeof(Object*));
```
We are copying `Object*`. Neither left nor the right side are `CORINFO_OBJECT_HANDLE`.

(`*(Object**)(...)` would be fine too instead of the `memcpy` ceremony here.)

Reply:
> (*(Object**)(...) would be fine too instead of the memcpy ceremony here.)

@jakobbotsch taught me to avoid such *unaligned* dereferences due to C++'s ability to optimize them as UB 🙂

Reply:
We learned that lesson in #67669, but whether or not it is UB depends on whether the runtime builds without strict aliasing. The JIT does build with strict aliasing, so this kind of type punning is UB in the JIT.

Reply:
There is no type aliasing in this line. The thing that you are copying is `Object*` and it is properly aligned.

Reply:
Ah, right, simplified then